compress.cc

```diff
@@ -133,6 +135,54 @@ static size_t read_uncompressed_size(const int8_t* in, size_t size) {
     return res;
 }
 
+size_t compressor::uncompress(managed_bytes_view in, bytes_ostream& os) const {
+    // default. fallback to linearized.
+    if (!in.is_linearized()) {
+        auto tmp = in.linearize();
+        return uncompress(managed_bytes_view(tmp), os);
+    }
+
+    auto view = in.current_fragment();
+    auto uncompressed_size = read_uncompressed_size(view.data(), view.size());
+
+    if (!_writes_uncompressed_size) {
+        view.remove_prefix(uncompressed_size_marker_size);
+    }
+
+    auto out = os.write_place_holder(uncompressed_size);
+    auto res = uncompress(reinterpret_cast<const char*>(view.data()), view.size(), reinterpret_cast<char*>(out), uncompressed_size);
+
+    assert(res == uncompressed_size);
+    return res;
+}
+
+size_t compressor::compress(managed_bytes_view in, bytes_ostream& os) const {
+    if (!in.is_linearized()) {
+        auto tmp = in.linearize();
+        return compress(managed_bytes_view(tmp), os);
+    }
+    auto view = in.current_fragment();
+    auto input_len = view.size();
+    if (view.size() > std::numeric_limits<uint32_t>::max()) {
+        throw std::invalid_argument("Data overflow");
+    }
+
+    auto size = compress_max_size(input_len);
+    if (!_writes_uncompressed_size) {
+        size += uncompressed_size_marker_size;
+    }
+    auto out = os.write_place_holder(size);
+    if (!_writes_uncompressed_size) {
+        auto n = write_uncompressed_size(out, size, input_len);
+        out += n;
+        size -= n;
+    }
+    auto actual = compress(reinterpret_cast<const char*>(view.data()), input_len, reinterpret_cast<char*>(out), size);
+    assert(actual <= size);
```

Comment:
`on_internal_error` instead of `assert`

Reply:
See comment above. Not saying you are factually wrong as scylla is today, just that it should not be as such.